File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/PBKDF2Params.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bccrypto.pbkdf2HmacSha256
import com.blockchaincommons.bccrypto.pbkdf2HmacSha512
import com.blockchaincommons.dcbor.Cbor

/**
 * PBKDF2-based key derivation parameters.
 *
 * CDDL:
 * ```
 * PBKDF2Params = [1, Salt, iterations: uint, HashType]
 * ```
 *
 * @property salt the salt used for key derivation
 * @property iterations the number of PBKDF2 iterations
 * @property hashType the hash algorithm (SHA-256 or SHA-512)
 */
class PBKDF2Params(
    val salt: Salt,
    val iterations: Int = DEFAULT_ITERATIONS,
    val hashType: HashType = HashType.SHA256,
) : KeyDerivation {

    /** Creates default parameters with a random 16-byte salt, 100 000 iterations, and SHA-256. */
    constructor() : this(Salt.createWithLength(SALT_LEN), DEFAULT_ITERATIONS, HashType.SHA256)

    override fun lock(contentKey: SymmetricKey, secret: ByteArray): EncryptedMessage {
        val derivedKey = deriveKey(secret)
        val encodedMethod = toCbor().toCborData()
        return derivedKey.encrypt(contentKey.data(), encodedMethod, null)
    }

    override fun unlock(encryptedMessage: EncryptedMessage, secret: ByteArray): SymmetricKey {
        val derivedKey = deriveKey(secret)
        val decrypted = derivedKey.decrypt(encryptedMessage)
        return SymmetricKey.fromData(decrypted)
    }

    private fun deriveKey(secret: ByteArray): SymmetricKey {
        val derived = when (hashType) {
            HashType.SHA256 -> pbkdf2HmacSha256(secret, salt.asBytes(), iterations, 32)
            HashType.SHA512 -> pbkdf2HmacSha512(secret, salt.asBytes(), iterations, 32)
        }
        return SymmetricKey.fromData(derived)
    }

    /** Encodes these parameters as a CBOR array. */
    fun toCbor(): Cbor = Cbor.fromArray(listOf(
        Cbor.fromInt(KeyDerivationMethod.PBKDF2.index),
        salt.untaggedCbor(),
        Cbor.fromInt(iterations),
        hashType.toCbor(),
    ))

    override fun toString(): String = "PBKDF2($hashType)"

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is PBKDF2Params) return false
        return salt == other.salt &&
            iterations == other.iterations &&
            hashType == other.hashType
    }

    override fun hashCode(): Int {
        var result = salt.hashCode()
        result = 31 * result + iterations
        result = 31 * result + hashType.hashCode()
        return result
    }

    companion object {
        const val SALT_LEN = 16
        const val DEFAULT_ITERATIONS = 100_000

        /**
         * Decodes [PBKDF2Params] from a CBOR array.
         *
         * The expected format is `[1, salt_bytes, iterations, hash_type_int]`.
         */
        fun fromCbor(cbor: Cbor): PBKDF2Params {
            val a = cbor.tryArray()
            if (a.size != 4) {
                throw BcComponentsException.general("Invalid PBKDF2Params: expected 4 elements, got ${a.size}")
            }
            // a[0] is the method index (already consumed by caller)
            val salt = Salt.fromUntaggedCbor(a[1])
            val iterations = a[2].tryInt()
            val hashType = HashType.fromCbor(a[3])
            return PBKDF2Params(salt, iterations, hashType)
        }
    }
}

File: csharp/BCComponents/BCComponents/PBKDF2Params.cs
using BlockchainCommons.BCCrypto;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// PBKDF2-based key derivation parameters.
/// </summary>
/// <remarks>
/// CDDL:
/// <code>
/// PBKDF2Params = [1, Salt, iterations: uint, HashType]
/// </code>
/// </remarks>
public sealed class PBKDF2Params : IEquatable<PBKDF2Params>, IKeyDerivation
{
    /// <summary>Default salt length in bytes.</summary>
    public const int SaltLen = 16;

    /// <summary>Default number of PBKDF2 iterations.</summary>
    public const int DefaultIterations = 100_000;

    /// <summary>Gets the salt used for key derivation.</summary>
    public Salt Salt { get; }

    /// <summary>Gets the number of PBKDF2 iterations.</summary>
    public int Iterations { get; }

    /// <summary>Gets the hash algorithm.</summary>
    public HashType HashType { get; }

    /// <summary>
    /// Creates PBKDF2 parameters with the given values.
    /// </summary>
    /// <param name="salt">The salt for key derivation.</param>
    /// <param name="iterations">The number of iterations.</param>
    /// <param name="hashType">The hash algorithm (default: SHA-256).</param>
    public PBKDF2Params(Salt salt, int iterations = DefaultIterations, HashType hashType = HashType.SHA256)
    {
        Salt = salt;
        Iterations = iterations;
        HashType = hashType;
    }

    /// <summary>
    /// Creates default PBKDF2 parameters with a random 16-byte salt, 100,000 iterations,
    /// and SHA-256.
    /// </summary>
    public PBKDF2Params()
        : this(Salt.CreateWithLength(SaltLen), DefaultIterations, HashType.SHA256)
    {
    }

    /// <inheritdoc/>
    public EncryptedMessage Lock(SymmetricKey contentKey, byte[] secret)
    {
        var derivedKey = DeriveKey(secret);
        var encodedMethod = ToCbor().ToCborData();
        return derivedKey.Encrypt(contentKey.Data, encodedMethod, null);
    }

    /// <inheritdoc/>
    public SymmetricKey Unlock(EncryptedMessage encryptedMessage, byte[] secret)
    {
        var derivedKey = DeriveKey(secret);
        var decrypted = derivedKey.Decrypt(encryptedMessage);
        return SymmetricKey.FromData(decrypted);
    }

    private SymmetricKey DeriveKey(byte[] secret)
    {
        var derived = HashType switch
        {
            HashType.SHA256 => Hash.Pbkdf2HmacSha256(secret, Salt.AsBytes(), (uint)Iterations, 32),
            HashType.SHA512 => Hash.Pbkdf2HmacSha512(secret, Salt.AsBytes(), (uint)Iterations, 32),
            _ => throw BCComponentsException.General($"Unknown hash type: {HashType}"),
        };
        return SymmetricKey.FromData(derived);
    }

    /// <summary>Encodes these parameters as a CBOR array.</summary>
    public Cbor ToCbor() => Cbor.FromList(new List<Cbor>
    {
        Cbor.FromInt((int)KeyDerivationMethod.PBKDF2),
        Salt.UntaggedCbor(),
        Cbor.FromInt(Iterations),
        HashType.ToCbor(),
    });

    /// <summary>Decodes <see cref="PBKDF2Params"/> from a CBOR array.</summary>
    /// <param name="cbor">The CBOR array value.</param>
    /// <returns>A new <see cref="PBKDF2Params"/>.</returns>
    public static PBKDF2Params FromCbor(Cbor cbor)
    {
        var a = cbor.TryIntoArray();
        if (a.Count != 4)
            throw BCComponentsException.General($"Invalid PBKDF2Params: expected 4 elements, got {a.Count}");
        var salt = Salt.FromUntaggedCbor(a[1]);
        var iterations = (int)a[2].TryIntoUInt64();
        var hashType = HashTypeExtensions.FromCbor(a[3]);
        return new PBKDF2Params(salt, iterations, hashType);
    }

    /// <inheritdoc/>
    public override string ToString() => $"PBKDF2({HashType})";

    /// <inheritdoc/>
    public bool Equals(PBKDF2Params? other)
    {
        if (other is null) return false;
        return Salt == other.Salt
            && Iterations == other.Iterations
            && HashType == other.HashType;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is PBKDF2Params p && Equals(p);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(Salt, Iterations, HashType);
}
